build libiceoryx_platform.so error for QNX

## Required information

**Operating system:**
E.g. Ubuntu 20.04 LTS

**Compiler version:**
aarch64-unknown-nto-qnx7.0.0-g++

**Observed result or behaviour:**

CMakeFiles/iceoryx_platform.dir/platform/qnx/source/socket.cpp.o: In function `iox_bind(int, sockaddr const*, unsigned int)':
socket.cpp:(.text+0x0): undefined reference to `bind'
CMakeFiles/iceoryx_platform.dir/platform/qnx/source/socket.cpp.o: In function `iox_socket(int, int, int)':
socket.cpp:(.text+0x8): undefined reference to `socket'
CMakeFiles/iceoryx_platform.dir/platform/qnx/source/socket.cpp.o: In function `iox_setsockopt(int, int, int, void const*, unsigned int)':
socket.cpp:(.text+0x10): undefined reference to `setsockopt'
CMakeFiles/iceoryx_platform.dir/platform/qnx/source/socket.cpp.o: In function `iox_sendto(int, void const*, unsigned long, int, sockaddr const*, unsigned int)':
socket.cpp:(.text+0x18): undefined reference to `sendto'
CMakeFiles/iceoryx_platform.dir/platform/qnx/source/socket.cpp.o: In function `iox_recvfrom(int, void*, unsigned long, int, sockaddr*, unsigned int*)':
socket.cpp:(.text+0x20): undefined reference to `recvfrom'
CMakeFiles/iceoryx_platform.dir/platform/qnx/source/socket.cpp.o: In function `iox_connect(int, sockaddr const*, unsigned int)':
socket.cpp:(.text+0x28): undefined reference to `connect'
collect2: error: ld returned 1 exit status
hoofs/CMakeFiles/iceoryx_platform.dir/build.make:173: recipe for target 'hoofs/libiceoryx_platform.so' failed
make[3]: *** [hoofs/libiceoryx_platform.so] Error 1
make[3]: Leaving directory '/home/qianchunlei/project/iceoryx/build'
CMakeFiles/Makefile2:165: recipe for target 'hoofs/CMakeFiles/iceoryx_platform.dir/all' failed
make[2]: *** [hoofs/CMakeFiles/iceoryx_platform.dir/all] Error 2
make[2]: Leaving directory '/home/qianchunlei/project/iceoryx/build'
CMakeFiles/Makefile2:172: recipe for target 'hoofs/CMakeFiles/iceoryx_platform.dir/rule' failed
make[1]: *** [hoofs/CMakeFiles/iceoryx_platform.dir/rule] Error 2
make[1]: Leaving directory '/home/qianchunlei/project/iceoryx/build'
Makefile:186: recipe for target 'iceoryx_platform' failed
make: *** [iceoryx_platform] Error 2


**Expected result or behaviour:**
no error

**Conditions where it occurred / Performed steps:**
build libiceoryx_platform.so on QNX

it looks like, link to libsocket is necessary under line https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_hoofs/CMakeLists.txt#L233
target_link_libraries(iceoryx_platform
        PRIVATE
        socket
        ${ICEORYX_SANITIZER_FLAGS})

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -34,6 +34,7 @@
 - Fix a few misspellings in log messages [\#962](https://github.com/eclipse-iceoryx/iceoryx/issues/962) thanks to @roehling
 - Fix typos in goals/non-goals document [\#968](https://github.com/eclipse-iceoryx/iceoryx/issues/968)
 - Catch deserialization errors for enums in publisher and subscriber options [\#989](https://github.com/eclipse-iceoryx/iceoryx/issues/989)
+- Fix linker error on QNX [\#1013](https://github.com/eclipse-iceoryx/iceoryx/issues/1013)
 
 **Refactoring:**
 

--- a/iceoryx_hoofs/CMakeLists.txt
+++ b/iceoryx_hoofs/CMakeLists.txt
@@ -104,8 +104,11 @@ if(GTest_FOUND) # only GTest_FOUND, just in case someone want's to use iceoryx_h
         GTest::gtest
         GTest::gmock
         ${CMAKE_DL_LIBS}
-        $<$<PLATFORM_ID:Linux>: rt>
     )
+
+    if(LINUX)
+        target_link_libraries(iceoryx_hoofs_testing PRIVATE rt)
+    endif()
 
     target_compile_options(iceoryx_hoofs_testing PRIVATE ${ICEORYX_WARNINGS})
 
@@ -194,8 +197,11 @@ target_link_libraries(iceoryx_hoofs
     iceoryx_hoofs::iceoryx_platform
     PRIVATE
     ${ICEORYX_SANITIZER_FLAGS}
-    $<$<PLATFORM_ID:Linux>: acl atomic ${CODE_COVERAGE_LIBS}>
 )
+
+if(LINUX)
+    target_link_libraries(iceoryx_hoofs PRIVATE acl atomic ${CODE_COVERAGE_LIBS})
+endif()
 
 target_compile_options(iceoryx_hoofs PRIVATE ${ICEORYX_WARNINGS} ${ICEORYX_SANITIZER_FLAGS})
 

--- a/iceoryx_hoofs/CMakeLists.txt
+++ b/iceoryx_hoofs/CMakeLists.txt
@@ -1,5 +1,5 @@
 # Copyright (c) 2019 - 2021 by Robert Bosch GmbH. All rights reserved.
-# Copyright (c) 2020 - 2021 by Apex.AI Inc. All rights reserved.
+# Copyright (c) 2020 - 2022 by Apex.AI Inc. All rights reserved.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
@@ -104,13 +104,10 @@ if(GTest_FOUND) # only GTest_FOUND, just in case someone want's to use iceoryx_h
         GTest::gtest
         GTest::gmock
         ${CMAKE_DL_LIBS}
+        $<$<PLATFORM_ID:Linux>: rt>
     )
 
     target_compile_options(iceoryx_hoofs_testing PRIVATE ${ICEORYX_WARNINGS})
-
-    if(LINUX)
-        target_link_libraries(iceoryx_hoofs_testing PRIVATE rt)
-    endif()
 
     setup_install_directories_and_export_package(
         TARGETS iceoryx_hoofs_testing
@@ -183,68 +180,22 @@ set_target_properties(iceoryx_hoofs PROPERTIES
     RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}"
 )
 
-if(LINUX)
-    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/platform/linux/)
-    target_link_libraries(iceoryx_hoofs PRIVATE acl atomic ${CODE_COVERAGE_LIBS})
-elseif(QNX)
-    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/platform/qnx/)
-elseif(APPLE)
-    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/platform/mac/)
-elseif(WIN32)
-    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/platform/win/)
-else()
-    message(WARNING "Could not detect supported platform, but I'm feeling lucky today." )
-endif()
-
-# this must be recurse since every platform has potentially different cpp files
-file ( GLOB_RECURSE ICEORYX_PLATFORM_FILES
-    ${ICEORYX_PLATFORM}/source/*.cpp
-)
-add_library(iceoryx_platform ${ICEORYX_PLATFORM_FILES})
-target_include_directories(iceoryx_platform
-    PUBLIC
-    $<BUILD_INTERFACE:${ICEORYX_PLATFORM}/include/>
-    $<INSTALL_INTERFACE:include/${PREFIX}>
-)
-add_library(iceoryx_hoofs::iceoryx_platform ALIAS iceoryx_platform)
-set_target_properties(iceoryx_platform PROPERTIES
-    CXX_STANDARD_REQUIRED ON
-    CXX_STANDARD ${ICEORYX_CXX_STANDARD}
-    POSITION_INDEPENDENT_CODE ON
-)
-
-target_link_libraries(iceoryx_platform PRIVATE ${ICEORYX_SANITIZER_FLAGS})
-
-target_compile_options(iceoryx_platform PRIVATE ${ICEORYX_WARNINGS} ${ICEORYX_SANITIZER_FLAGS})
-
-if(LINUX)
-    target_link_libraries(iceoryx_platform
-        PUBLIC
-        rt
-        pthread
-    )
-endif(LINUX)
-
 if(PERFORM_CLANG_TIDY)
     set_target_properties(
         iceoryx_hoofs PROPERTIES CXX_CLANG_TIDY "${PERFORM_CLANG_TIDY}"
     )
 endif(PERFORM_CLANG_TIDY)
 
-if(QNX)
-    target_link_libraries(iceoryx_hoofs
-        PUBLIC
-        iceoryx_hoofs::iceoryx_platform
-        PRIVATE
-        socket
-        ${ICEORYX_SANITIZER_FLAGS})
-else()
-    target_link_libraries(iceoryx_hoofs
-        PUBLIC
-        iceoryx_hoofs::iceoryx_platform
-        PRIVATE
-        ${ICEORYX_SANITIZER_FLAGS})
-endif(QNX)
+add_subdirectory(platform)
+add_library(iceoryx_hoofs::iceoryx_platform ALIAS iceoryx_platform)
+
+target_link_libraries(iceoryx_hoofs
+    PUBLIC
+    iceoryx_hoofs::iceoryx_platform
+    PRIVATE
+    ${ICEORYX_SANITIZER_FLAGS}
+    $<$<PLATFORM_ID:Linux>: acl atomic ${CODE_COVERAGE_LIBS}>
+)
 
 target_compile_options(iceoryx_hoofs PRIVATE ${ICEORYX_WARNINGS} ${ICEORYX_SANITIZER_FLAGS})
 
@@ -254,8 +205,6 @@ target_include_directories(iceoryx_hoofs
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
     $<INSTALL_INTERFACE:include/${PREFIX}>
 )
-
-
 
 #
 ########## exporting library ##########
@@ -267,6 +216,7 @@ install(
     DESTINATION include/${PREFIX}
     COMPONENT dev
 )
+
 setup_install_directories_and_export_package(
     TARGETS iceoryx_hoofs iceoryx_platform
     INCLUDE_DIRECTORY include/

--- a/iceoryx_hoofs/platform/CMakeLists.txt
+++ b/iceoryx_hoofs/platform/CMakeLists.txt
@@ -1,0 +1,65 @@
+# Copyright (c) 2019 - 2021 by Robert Bosch GmbH. All rights reserved.
+# Copyright (c) 2020 - 2022 by Apex.AI Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.5)
+
+if(LINUX)
+    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/linux/)
+elseif(QNX)
+    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/qnx/)
+elseif(APPLE)
+    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/mac/)
+elseif(WIN32)
+    set(ICEORYX_PLATFORM ${CMAKE_CURRENT_SOURCE_DIR}/win/)
+else()
+    message(WARNING "Could not detect supported platform, but I'm feeling lucky today." )
+endif()
+
+set(ICEORYX_PLATFORM ${ICEORYX_PLATFORM} CACHE PATH "" FORCE)
+
+# this must be recurse since every platform has potentially different cpp files
+file ( GLOB_RECURSE ICEORYX_PLATFORM_FILES
+    ${ICEORYX_PLATFORM}/source/*.cpp
+)
+add_library(iceoryx_platform ${ICEORYX_PLATFORM_FILES})
+target_include_directories(iceoryx_platform
+    PUBLIC
+    $<BUILD_INTERFACE:${ICEORYX_PLATFORM}/include/>
+    $<INSTALL_INTERFACE:include/${PREFIX}>
+)
+set_target_properties(iceoryx_platform PROPERTIES
+    CXX_STANDARD_REQUIRED ON
+    CXX_STANDARD ${ICEORYX_CXX_STANDARD}
+    POSITION_INDEPENDENT_CODE ON
+)
+
+target_link_libraries(iceoryx_platform PRIVATE ${ICEORYX_SANITIZER_FLAGS})
+
+target_compile_options(iceoryx_platform PRIVATE ${ICEORYX_WARNINGS} ${ICEORYX_SANITIZER_FLAGS})
+
+if(LINUX)
+    target_link_libraries(iceoryx_platform
+        PUBLIC
+        rt
+        pthread
+    )
+elseif(QNX)
+    target_link_libraries(iceoryx_platform
+        PRIVATE
+        socket
+    )
+endif()

